### PR TITLE
:seedling: Remove CI image pull and tagging for kubeadm injection script

### PR DESCRIPTION
**Pull request by jsturtevant**
**What this PR does / why we need it**:
After the switch from k8s.gcr.io to registry.k8s.io (see https://github.com/kubernetes/kubernetes/pull/109938 for details), the CI jobs are outputting the following when trying to extract the ci images:

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/capz-conformance-master/1532465763934277632

```
[2022-06-02 21:09:45] 
[2022-06-02 21:09:49] unpacking registry.k8s.io/kube-apiserver-amd64:v1.25.0-alpha.0.734_ba502ee555924a (sha256:4600d5b413457521bd7c01d24c0b21288c63f05137aa7547b6e0cb5e9a654bef)...done
[2022-06-02 21:09:49] ctr: image "k8s.gcr.io/kube-apiserver-amd64:v1.25.0-alpha.0.734_ba502ee555924a": not found`
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2351

### other info
I found [our jobs](https://testgrid.k8s.io/provider-azure-master-signal#capz-conformance) are not failing since kubeadm is smart enough because we give it the `kubernetesversion` in the job config, even without this to pull the correct images so our jobs are still passing.  

See https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/9e793b42f79b65cbc830e711dce63a2c1e7c82c0/templates/test/ci/cluster-template-prow-ci-version-windows-containerd-2022.yaml#L90

Is there a reason to pull these images?

### CODE REVIEW

**Comment by jackfrancis** [comment]
lgtm

**Comment by sbueringer** [comment]
> Is there a reason to pull these images?

@jsturtevant We needed the wget + import + tag of those images as they were not available in a registry (at least for CAPO and CAPA too afaik). 

It's interesting that for CAPZ it works with image pull, e.g. here is a successfull image pull: 
* https://storage.googleapis.com/kubernetes-jenkins/logs/capz-conformance-master/1532465763934277632/artifacts/clusters/capz-conf-x78nve/machines/capz-conf-x78nve-control-plane-gsp7w/containerd.log. 
> Jun 02 21:09:50.735341 capz-conf-x78nve-control-plane-cxrb9 containerd[1319]: time="2022-06-02T21:09:50.735271164Z" level=info msg="PullImage \"gcr.io/k8s-staging-ci-images/kube-apiserver:v1.25.0-alpha.0.734_ba502ee555924a\""
> Jun 02 21:09:54.684557 capz-conf-x78nve-control-plane-cxrb9 containerd[1319]: time="2022-06-02T21:09:54.684487533Z" level=info msg="ImageCreate event &ImageCreate{Name:gcr.io/k8s-staging-ci-images/kube-apiserver:v1.25.0-alpha.0.734_ba502ee555924a,Labels:map[string]string{io.cri-containerd.image: managed,},XXX_unrecognized:[],}"

Note: The download script fails in CAPZ but this doesn't fail the entire bootstrap process.

Looks like the CAPO tests are currently failing since the change in image names:
* https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-openstack#periodic-conformance-test-main
* https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-provider-openstack-conformance-test-main-with-k8s-ci-artifacts/1531787342853967872/artifacts/clusters/bootstrap/instances/conformance-6zyam3/cluster-conformance-6zyam3-control-plane-zzknf/cloud-final.log
> ctr: image "k8s.gcr.io/kube-apiserver-amd64:v1.25.0-alpha.0.729_62d9f8ba80f4cc": not found
> ...
>  [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-apiserver:v1.25.0-alpha.0.729_62d9f8ba80f4cc

In general the change looks fine, but I think we have to use the old/new registry depending on the Kubernetes version, otherwise jobs using <v1.25 will fail (if there are any).


@dims Do you know why those images might be available for CAPZ but not for CAPO? (I read about the cloud-local redirects, but I don't know if that explains it)

**Comment by sbueringer** [comment]
cc @mdbooth (for CAPO)

**Comment by dims** [comment]
@ameukam @BenTheElder can you please peek?

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Comment by vincepri** [comment]
/hold

for the above

**Comment by jsturtevant** [comment]
>@jsturtevant We needed the wget + import + tag of those images as they were not available in a registry (at least for CAPO and CAPA too afaik).

something might be wrong in the configuration on openstack.  Kubeadm in Azure is pulling: `gcr.io/k8s-staging-ci-images/kube-proxy:v1.25.0-alpha.0.859_198dd7668a19c8\` where as the kubeadm in Openstack is pulling `k8s.gcr.io/kube-controller-manager:v1.25.0-alpha.0.729_62d9f8ba80f4cc`

notice the difference between `gcr.io/k8s-staging-ci-images` and `k8s.gcr.io/kube-controller-manager`. I don't think the  CI images are pushed to `k8s.gcr.io` but are pushed to the `gcr.io/k8s-staging-ci-images` staging repo?  I am not sure why kubeadm is configured to pull from there in Openstack and not Azure.

**Comment by BenTheElder** [comment]
Yeah,
> ctr: image "k8s.gcr.io/kube-apiserver-amd64:v1.25.0-alpha.0.729_62d9f8ba80f4cc": not found

An image with a tag like this won't be in k8s.gcr.io, tags there have to be promoted in the k8s.io repo, only release tags are present.

**Comment by sbueringer** [comment]
Thx! Sorry should have seen that.

Okay looks like the following:
* we have to download but not the images
* kubeadm automatically uses the CI registry if kubeadm is a CI binary: https://github.com/kubernetes/kubernetes/blob/703f2a7b86a2340fce92b3f9dae12a30254620a3/cmd/kubeadm/app/util/config/common.go#L94-L96
* In CAPZ the KCP image repo is not set so kubeadm uses the CI registry: https://storage.googleapis.com/kubernetes-jenkins/logs/capz-conformance-master/1534363744480530432/artifacts/clusters/bootstrap/resources/capz-conf-bl9jla/KubeadmControlPlane/capz-conf-bl9jla-control-plane.yaml
* In CAPO the KCP image repo is pinned to `k8s.gcr.io`: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-provider-openstack-conformance-test-main-with-k8s-ci-artifacts/1534363241298268160/artifacts/clusters/bootstrap/resources/conformance-8es4na/KubeadmControlPlane/cluster-conformance-8es4na-control-plane.yaml

I would suggest the following:
* let's drop the entire image download + tagging from the script because:
    * we don't need it
    * we can avoid adding a Kubernetes version dependent if/else to tag the old/new image name based on Kubernetes version
* CAPO should not pin to the production registry

**Comment by sbueringer** [comment]
Looping in CAPA in case they are (still) using this script cc @pydctw @richardcase

**Comment by sbueringer** [comment]
/lgtm

It would be good to wait until next week with merge. Just to get feedback from CAPO/CAPA (one of the CAPO maintainers at least is out this week)

**Comment by jsturtevant** [comment]
/retitle Remove CI image pull and tagging for kubeadm injection script

**Review by fabriziopandini — COMMENTED** [review]
As per comments in the thread above
/lgtm
/approve

**Comment by sedefsavas** [comment]
[CAPA k8s-main conformance](https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-aws#periodic-conformance-main-k8s-main&show-stale-tests=) test is broken since https://github.com/kubernetes/kubernetes/pull/109938 is merged, not sure how it works for CAPZ but we were using the pulled images here. 

 I will need to further test if just changing the registry here works instead of removing it.

**Comment by sedefsavas** [comment]
For unblocking CAPZ, how about just changing k8s.gcr.io to registry.k8s.io in this PR so that CAPZ script does not fail? @jsturtevant @CecileRobertMichon

**Comment by sedefsavas** [comment]
> This is what I did in the initial version of this PR but it was discover that the logic for pull/tagging images is not required so switch to removing it https://github.com/kubernetes-sigs/cluster-api/pull/6590#issuecomment-1149463509

CAPA also doesn't pin the registry to k8s.gcr.io but still it stopped working for us, need to collect the logs.

**Comment by sedefsavas** [comment]
@sbueringer follow up on https://github.com/kubernetes-sigs/cluster-api/pull/6590#issuecomment-1149463509 
> `let's drop the entire image download + tagging from the script`

Without importing and tagging part, ci kubeadm version fails to pull images. Errors I see:

```
root@ip-10-0-149-177:/var/snap/amazon-ssm-agent/5656# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.0-alpha.0.956+0fe0dbf3fb8cf5", GitCommit:"0fe0dbf3fb8cf501c24c87f4113a3819cb86a550", GitTreeState:"clean", BuildDate:"2022-06-13T17:30:11Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}

root@ip-10-0-149-177:/var/snap/amazon-ssm-agent/5656# kubeadm init --config /run/kubeadm/kubeadm.yaml 
[init] Using Kubernetes version: v1.25.0-alpha.0.956+0fe0dbf3fb8cf5
[preflight] Running pre-flight checks
[preflight] Pulling images required for setting up a Kubernetes cluster
[preflight] This might take a minute or two, depending on the speed of your internet connection
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
error execution phase preflight: [preflight] Some fatal errors occurred:
        [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-apiserver:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: output: time="2022-06-13T21:58:02Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-apiserver:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": failed to resolve reference \"registry.k8s.io/kube-apiserver:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": registry.k8s.io/kube-apiserver:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: not found"
, error: exit status 1
        [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-controller-manager:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: output: time="2022-06-13T21:58:04Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-controller-manager:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": failed to resolve reference \"registry.k8s.io/kube-controller-manager:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": registry.k8s.io/kube-controller-manager:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: not found"
, error: exit status 1
        [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-scheduler:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: output: time="2022-06-13T21:58:05Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-scheduler:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": failed to resolve reference \"registry.k8s.io/kube-scheduler:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": registry.k8s.io/kube-scheduler:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: not found"
, error: exit status 1
        [ERROR ImagePull]: failed to pull image registry.k8s.io/kube-proxy:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: output: time="2022-06-13T21:58:06Z" level=fatal msg="pulling image: rpc error: code = NotFound desc = failed to pull and unpack image \"registry.k8s.io/kube-proxy:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": failed to resolve reference \"registry.k8s.io/kube-proxy:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5\": registry.k8s.io/kube-proxy:v1.25.0-alpha.0.956_0fe0dbf3fb8cf5: not found"
, error: exit status 1
[preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
To see the stack trace of this error execute with --v=5 or higher

```


After changing the registry to the new one in this PR, it works. So looks like we still need to import and tag images.

I don't know how CAPZ test does not fail with this kubeadm error.

**Comment by sbueringer** [comment]
Looks like there is something wrong with kubeadm somehow.

It works in CAPZ because kubeadm detects the CI version and it pulls images from the CI registry automatically, e.g. `gcr.io/k8s-staging-ci-images/kube-apiserver:v1.25.0-alpha.0.965_c0cc9116677732` (because of https://github.com/kubernetes/kubernetes/blob/703f2a7b86a2340fce92b3f9dae12a30254620a3/cmd/kubeadm/app/util/config/common.go#L94-L96)

Looks like in CAPA kubeadm somehow tries to pull images from the (new) production registry (registry.k8s.io).

**Comment by sbueringer** [comment]
I think I found it. CAPZ sets a `ci/` version in clusterConfiguration.kubernetesVersion
* https://storage.googleapis.com/kubernetes-jenkins/logs/capz-conformance-master/1532465763934277632/artifacts/clusters/bootstrap/resources/capz-conf-x78nve/KubeadmControlPlane/capz-conf-x78nve-control-plane.yaml
  > kubernetesVersion: ci/v1.25.0-alpha.0.734+ba502ee555924a

kubeadm is looking for the ci prefix there, then it automatically uses the test registry/imageRepository (gcr.io/k8s-staging-ci-images)

**Comment by CecileRobertMichon** [comment]
Good find @sbueringer. @sedefsavas can CAPA add the ci/ prefix too? Then we don't need additional script lines to pull the images and it works for all versions.

CAPO should also do the same.

**Comment by sedefsavas** [comment]
Good find!! We will change it in CAPA.

/lgtm

**Comment by sbueringer** [comment]
Okay I think we can go ahead and merge it.

Thx everyone!

/hold cancel

**Pull request merged**
